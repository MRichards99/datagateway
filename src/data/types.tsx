export interface InvestigationData {
  ID: string;
  TITLE: string;
  VISIT_ID: number;
  RB_NUMBER: string;
  DOI: string;
  SIZE: number;
  INSTRUMENT: { NAME: string };
  STARTDATE: Date;
  ENDDATE: Date;
  [key: string]: string | number | Date | { NAME: string };
}

export interface DatasetData {
  ID: string;
  NAME: string;
  SIZE: number;
  CREATE_TIME: Date;
  MOD_TIME: Date;
  [key: string]: string | number | Date;
}

export interface DatafileData {
  ID: string;
  NAME: string;
  LOCATION: string;
  SIZE: number;
  MOD_TIME: Date;
  [key: string]: string | number | Date;
}

export interface FoodData {
  id: number;
  dessert: string;
  calories: number;
  fat: number;
  carbs: number;
  protein: number;
  [key: string]: string | number;
}

export type EntityType =
  | InvestigationData
  | DatasetData
  | DatafileData
  | FoodData;
