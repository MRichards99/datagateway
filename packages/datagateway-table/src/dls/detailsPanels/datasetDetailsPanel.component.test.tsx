import React from 'react';
import { createShallow, createMount } from '@material-ui/core/test-utils';
import DatasetsDetailsPanel from './datasetDetailsPanel.component';
import { Dataset } from '../../state/app.types';

describe('Dataset details panel component', () => {
  let shallow;
  let mount;
  let rowData: Dataset;
  const detailsPanelResize = jest.fn();
  const fetchDetails = jest.fn();

  beforeEach(() => {
    shallow = createShallow({ untilSelector: 'div' });
    mount = createMount();
    rowData = {
      ID: 1,
      NAME: 'Test 1',
      MOD_TIME: '2019-06-10',
      CREATE_TIME: '2019-06-11',
      INVESTIGATION_ID: 1,
    };
  });

  afterEach(() => {
    mount.cleanUp();
    detailsPanelResize.mockClear();
    fetchDetails.mockClear();
  });

  it('renders correctly', () => {
    const wrapper = shallow(
      <DatasetsDetailsPanel
        rowData={rowData}
        detailsPanelResize={detailsPanelResize}
        fetchDetails={fetchDetails}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders type tab when present in the data', () => {
    rowData.DATASETTYPE = {
      ID: 7,
      NAME: 'Test type',
      DESCRIPTION: 'Test type description',
    };

    const wrapper = shallow(
      <DatasetsDetailsPanel
        rowData={rowData}
        detailsPanelResize={detailsPanelResize}
        fetchDetails={fetchDetails}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders calculate size button when size has not been calculated', () => {
    const { SIZE, ...rowDataWithoutSize } = rowData;

    const wrapper = shallow(
      <DatasetsDetailsPanel
        rowData={rowDataWithoutSize}
        detailsPanelResize={detailsPanelResize}
        fetchDetails={fetchDetails}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('calculates size when button is clicked', () => {
    const { SIZE, ...rowDataWithoutSize } = rowData;

    const wrapper = mount(
      <DatasetsDetailsPanel
        rowData={rowDataWithoutSize}
        detailsPanelResize={detailsPanelResize}
        fetchDetails={fetchDetails}
      />
    );

    wrapper.find('#dataset-details-panel button').simulate('click');

    // TODO: test this calls some action that requests to getSize
    expect(true);
  });

  it('calls detailsPanelResize on load and when tabs are switched between', () => {
    rowData.DATASETTYPE = {
      ID: 7,
      NAME: 'Test type',
      DESCRIPTION: 'Test type description',
    };

    const wrapper = mount(
      <DatasetsDetailsPanel
        rowData={rowData}
        detailsPanelResize={detailsPanelResize}
        fetchDetails={fetchDetails}
      />
    );

    expect(detailsPanelResize).toHaveBeenCalledTimes(1);

    wrapper
      .find('#dataset-type-tab')
      .hostNodes()
      .simulate('click');

    expect(detailsPanelResize).toHaveBeenCalledTimes(2);
  });

  it('calls fetchDetails on load', () => {
    const wrapper = mount(
      <DatasetsDetailsPanel
        rowData={rowData}
        detailsPanelResize={detailsPanelResize}
        fetchDetails={fetchDetails}
      />
    );

    expect(fetchDetails).toHaveBeenCalled();
    expect(fetchDetails).toHaveBeenCalledWith(1);
  });
});
