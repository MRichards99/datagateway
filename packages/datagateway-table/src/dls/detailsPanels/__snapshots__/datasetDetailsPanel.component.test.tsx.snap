// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dataset details panel component renders calculate size button when size has not been calculated 1`] = `
<div>
  <WithStyles(ForwardRef(Tabs))
    aria-label="dataset-details-tabs"
    onChange={[Function]}
    value="details"
  >
    <WithStyles(ForwardRef(Tab))
      aria-controls="dataset-details-panel"
      id="dataset-details-tab"
      label="Dataset Details"
      value="details"
    />
  </WithStyles(ForwardRef(Tabs))>
  <div
    aria-labelledby="dataset-details-tab"
    hidden={false}
    id="dataset-details-panel"
    role="tabpanel"
  >
    <WithStyles(ForwardRef(Typography))
      variant="body2"
    >
      <b>
        Name:
      </b>
       
      Test 1
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      variant="body2"
    >
      <b>
        Description:
      </b>
       
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      variant="body2"
    >
      <b>
        Start Date:
      </b>
       
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      variant="body2"
    >
      <b>
        Description:
      </b>
       
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      variant="body2"
    >
      <b>
        Total File Size:
      </b>
       
      <WithStyles(ForwardRef(Button))
        color="primary"
        id="calculate-size-btn"
        onClick={[Function]}
        size="small"
        variant="outlined"
      >
        Calculate
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(Typography))>
  </div>
</div>
`;

exports[`Dataset details panel component renders correctly 1`] = `
<div>
  <WithStyles(ForwardRef(Tabs))
    aria-label="dataset-details-tabs"
    onChange={[Function]}
    value="details"
  >
    <WithStyles(ForwardRef(Tab))
      aria-controls="dataset-details-panel"
      id="dataset-details-tab"
      label="Dataset Details"
      value="details"
    />
  </WithStyles(ForwardRef(Tabs))>
  <div
    aria-labelledby="dataset-details-tab"
    hidden={false}
    id="dataset-details-panel"
    role="tabpanel"
  >
    <WithStyles(ForwardRef(Typography))
      variant="body2"
    >
      <b>
        Name:
      </b>
       
      Test 1
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      variant="body2"
    >
      <b>
        Description:
      </b>
       
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      variant="body2"
    >
      <b>
        Start Date:
      </b>
       
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      variant="body2"
    >
      <b>
        Description:
      </b>
       
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      variant="body2"
    >
      <b>
        Total File Size:
      </b>
       
      <WithStyles(ForwardRef(Button))
        color="primary"
        id="calculate-size-btn"
        onClick={[Function]}
        size="small"
        variant="outlined"
      >
        Calculate
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(Typography))>
  </div>
</div>
`;

exports[`Dataset details panel component renders type tab when present in the data 1`] = `
<div>
  <WithStyles(ForwardRef(Tabs))
    aria-label="dataset-details-tabs"
    onChange={[Function]}
    value="details"
  >
    <WithStyles(ForwardRef(Tab))
      aria-controls="dataset-details-panel"
      id="dataset-details-tab"
      label="Dataset Details"
      value="details"
    />
    <WithStyles(ForwardRef(Tab))
      aria-controls="dataset-type-panel"
      id="dataset-type-tab"
      label="Dataset Type"
      value="type"
    />
  </WithStyles(ForwardRef(Tabs))>
  <div
    aria-labelledby="dataset-details-tab"
    hidden={false}
    id="dataset-details-panel"
    role="tabpanel"
  >
    <WithStyles(ForwardRef(Typography))
      variant="body2"
    >
      <b>
        Name:
      </b>
       
      Test 1
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      variant="body2"
    >
      <b>
        Description:
      </b>
       
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      variant="body2"
    >
      <b>
        Start Date:
      </b>
       
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      variant="body2"
    >
      <b>
        Description:
      </b>
       
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      variant="body2"
    >
      <b>
        Total File Size:
      </b>
       
      <WithStyles(ForwardRef(Button))
        color="primary"
        id="calculate-size-btn"
        onClick={[Function]}
        size="small"
        variant="outlined"
      >
        Calculate
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(Typography))>
  </div>
  <div
    aria-labelledby="dataset-type-tab"
    hidden={true}
    id="dataset-type-panel"
    role="tabpanel"
  >
    <WithStyles(ForwardRef(Typography))
      variant="body2"
    >
      <b>
        Name:
      </b>
       
      Test type
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      variant="body2"
    >
      <b>
        Description:
      </b>
       
      Test type description
    </WithStyles(ForwardRef(Typography))>
  </div>
</div>
`;
